harness/determined/_preemption.py

```diff
@@ -0,0 +1,142 @@
+        assert self._wsock is not None
+
+        with api.get(
+            self._master_url,
+            f"/api/v1/trials/{self._trial_id}/signals/preemption",
+            stream=True,
+            cert=self._cert,
+        ) as r:
+            lines = r.iter_lines()
+
+            # The first response is always immediate, so we always block for it.  This is doubly
+            # important because we can only select on the HTTP connection socket _after_ we call
+            # next(lines) at least once.
+            for line in lines:
+                if not line:
+                    # Empty keepalive message.
+                    continue
+                break
+
+            j = json.loads(line)
+
+            with self._cond:
+                self._should_preempt = j["result"]["preempt"]
+                # Wake the main thread in case it was waiting for the main response.
+                self._cond.notify()
+
+            if self._should_preempt:
+                return
+
+            # It may be a long time between the first response and the second.  We want to be
+            # responsive to .close() calls from the main thread, so we block in input from either
+            # the HTTP connection socket or on our own _rsock socket, over which we can be alerted
+            # to .close() calls.
+            with selectors.DefaultSelector() as sel:
+                sel.register(r.raw, selectors.EVENT_READ)
+                sel.register(self._rsock, selectors.EVENT_READ)
+
+                while True:
+                    for key, _ in sel.select():
+                        if key.fileobj == self._rsock:
+                            # self._rsock is only written to when we are supposed to close.
+                            return
+
+                        assert key.fileobj == r.raw
+                        # Calling next() is blocking, but it should only block for a trivial amount
+                        # of time; basically only in the case that one message from the preemption
+                        # API got broken into two network packets.
+                        # r.raw._connection.sock.setblocking(True)
+                        line = next(lines)
+                        # r.raw._connection.sock.setblocking(False)
+
+                        if not line:
+                            # Empty keepalive message.
+                            continue
+
+                        j = json.loads(line)
+
+                        self._should_preempt = j["result"]["preempt"]
+
+                        if self._should_preempt:
+                            return
+
+    def start(self) -> None:
+        self._rsock, self._wsock = socket.socketpair()
+        super().start()
+
+    def close(self) -> None:
+        if self._wsock is None:
+            return
+        # Sending any message at all will wake up the selector loop.
+        self._wsock.send(b"quit")
+        self.join()
+        self._rsock = None
```

Comment:
should we close these raw sockets explicitly instead of relying on GC?

Reply:
doh.